### PR TITLE
load extensions through pythons entry-points machinery

### PR DESCRIPTION
**Pull request by tjfulle**
## What is it?

This PR adds the ability to load spack extensions through `importlib.metadata.entry_points` in addition to the regular configuration variable.

## Why?

This allows one to `pip install my-extension` and have spack find it without having to provide additional configuration.

## Notes

- See initial discussion at https://spackpm.slack.com/archives/C5W7NKZJT/p1706560654393159
- Documentation and tests have not been provided, but can if this PR is deemed useful and/or of interest.

## Usage

In your package's `pyproject.toml` add the following:

```toml
[project.entry-points."spack.extensions"]
"ext-name" = "module-name:function-that-returns-list-of-extension-paths"

[project.entry-points."spack.config"]
"ext-name" = "module-name:function-that-returns-config-scope-path"
```

## Example

`pyproject.toml`

```toml
[project]
name = "baz"
description = "baz"
version = "0.1.0"
requires-python = ">=3.9"

[project.entry-points."spack.extensions"]
baz = "baz:get_spack_extension_paths"

[project.entry-points."spack.config"]
baz = "baz:get_spack_config_dir"
```

`baz/__init__.py`:

```python
import importlib.resources

def get_spack_extension_paths() -> list[str]:
    dirname = importlib.resources.files("baz").joinpath("spack-baz")
    if dirname.exists():  # type: ignore
        return [str(dirname)]
    return []

def get_spack_config_dir() -> str:
    dirname = importlib.resources.files("baz").joinpath("spack-config")
    if dirname.exists():  # type: ignore
        return str(dirname)
    return []
```

### CODE REVIEW

**Comment by tjfulle** [comment]
@psakievich

**Comment by psakievich** [comment]
This looks valuable to me.  Personally, I'm okay with this feature being conditionally available if `SPACK_PYTHON>=3.8` but I think we need to make sure the imports are conditional on python version so we can support older python versions that spack still supports.

**Comment by tjfulle** [comment]
The last commit hides `importlib.metadata` (which is available in python >= 3.8) so that this PR is compatible with older python.  Older python will not be able to use the entry points as a means to inject user configurations into spack.

File: lib/spack/spack/config.py
Comment (alalazo) [inline lib/spack/spack/config.py]: I'm not super familiar with entry points. Can you explain when we'll get the "try" succeeding, and when we fall back on the exceptional case?

File: lib/spack/spack/config.py
Comment (alalazo) [inline lib/spack/spack/config.py]: (minor) Here I would just define:
```suggestion
def paths_from_entry_points() -> List[str]:
```
and return the list of paths from entry points

File: lib/spack/spack/config.py
Comment (alalazo) [inline lib/spack/spack/config.py]: (minor) While here I would:
```suggestion
    configuration_paths.extend(paths_from_entry_points())
```

File: lib/spack/spack/extensions.py
Comment (alalazo) [inline lib/spack/spack/extensions.py]: This code and the import above are duplicated. They can probably be factored in a separate function and used both here and in `spack.config`.

**Review by alalazo — COMMENTED** [review]
I like the idea, it might even blend well with #40773 and the other idea of being able to turn Spack extensions into Spack packages.

File: lib/spack/spack/config.py
Comment (tjfulle) [reply lib/spack/spack/config.py]: @alalazo - the exceptional case is for Python 3.8, the default case is for 3.9 and newer.

**Comment by tjfulle** [comment]
Thanks for the feedback @alalazo !

I addressed your comments and added documentation.  I'll likely need some help testing.  A test would require creating a python venv, installing a dummy package into it, then testing whether spack can read the entry points.  Or, that could all be mocked.  Not sure which is best...

**Comment by tjfulle** [comment]
@alalazo  and @psakievich : bump

**Review by psakievich — DISMISSED** [review]
This looks good to me. @alalazo thoughts on the changes?

**Comment by alalazo** [comment]
@psakievich I like the idea, but this PR needs to pass CI and add some unit tests before being merged. Also, we need to state in the docs what are the additional requests we make for people who want to use this feature (I guess either Python > 3.8 or `setuptools` installed).

**Comment by tjfulle** [comment]
thanks for the feedback @alalazo

- documents have been updated with a note about requiring `python>=3.8` 
- 2 new unit tests give nearly 100% coverage of new code added.  the only code that doesn't get covered is some defensive checks for `python 3.7` (which reached [EOL last year](https://devguide.python.org/versions/))

**Comment by alalazo** [comment]
@tjfulle Thanks! I'll review either today or beginning of next week.

**Comment by tjfulle** [comment]
@alalazo said:

> I like the idea, it might even blend well with https://github.com/spack/spack/pull/40773 and the other idea of being able to turn Spack extensions into Spack packages.

Incidentally, this is essentially what we are doing internally in my group at Sandia: `python -m pip install <my group's extensions>` and then Spack "just works" with our extensions, repos, configurations, etc.  If the extension were packaged as a `PythonPackage`, then just replace `python -m pip install` with `spack install`.  No further work needed and the Spack extensions are ready to roll.

File: lib/spack/spack/test/entry_points.py
Comment (alalazo) [inline lib/spack/spack/test/entry_points.py]: (minor) You can turn this into a fixture, maybe use `tmp_path` instead of `tmpdir` (deprecated):
```suggestion
@pytest.fixture()
def mock_entry_points(tmp_path, monkeypatch):
```

File: lib/spack/spack/test/entry_points.py
Comment (alalazo) [inline lib/spack/spack/test/entry_points.py]: Then this will look like (edited in web UI, so beware of typos):
```suggestion
def test_spack_entry_point_config(tmp_path, mock_entry_points):
    """Test config scope entry point"""    
```

File: lib/spack/spack/test/entry_points.py
Comment (alalazo) [inline lib/spack/spack/test/entry_points.py]: Since `tmp_path` is a `pathlib.Path` object:
```suggestion
    my_config_path = tmp_path / "spack" / "etc"
```

File: lib/spack/spack/test/entry_points.py
Comment (alalazo) [inline lib/spack/spack/test/entry_points.py]: Similarly here

**Review by alalazo — CHANGES_REQUESTED** [review]
Thanks! This LGTM, I just left a couple of minor comments for tests. Let me know if you agree with them. In the meanwhile I'll test drive the PR, but I expect to merge it shortly.

**Comment by alalazo** [comment]
@tjfulle Would you agree putting a note in the docs saying that the entry-point support is still experimental and we might change / improve the layout anytime? I'd like to merge this one soon, but I'm wondering if we can simplify the required layout:
```
my-package/
├── lib
│   ├── my_package
│   │   └── __init__.py
│   └── spack-scripting/  # the spack extensions
└── pyproject.toml
```

**Comment by tjfulle** [comment]
@alalazo - thanks for the feedback!  I updated the test and the library function to accept `pathlib.Path` objects.

**Comment by tjfulle** [comment]
@alalazo said:

> Would you agree putting a note in the docs saying that the entry-point support is still experimental and we might change / improve the layout anytime? I'd like to merge this one soon, but I'm wondering if we can simplify the required layout:

The layout in the documentation is the "`src`" layout recommended by the [python packaging authority](https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/).  I would recommend leaving it.  Furthermore, I don't think Spack should make a recommendation (outside of what [pypa](https://packaging.python.org/en/latest/) has already recommended) on a python package's layout.  It should just promise that if a package registers entry point hooks the meet the contract, spack will load them.  The user can use whatever layout they choose.

**Comment by tjfulle** [comment]
just noticed now that the package layout in the examples used `lib` instead of `src`.  It's fixed now.

**Comment by alalazo** [comment]
> The layout in the documentation is the "src" layout recommended by the [python packaging authority](https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/). [ .. ]

I agree with what you say, what bother me is more on Spack side - the necessity to have a `spack-xxx` subfolder within another package :thinking:[^1] In any case not a deal-breaker. We can try to simplify that later in case.

[^1]: This is mostly due to the fact that we didn't design extension with entry points in mind years ago.

**Review by alalazo — APPROVED** [review]
Thanks!

**Comment by haampie** [comment]
This has problems, reported on Slack.

After:

```
ImportError: cannot import name 'metadata'
ModuleNotFoundError: No module named 'importlib_metadata'
```

things can still go wrong because of

```
    import pkg_resources  # type: ignore
```

triggering

```
SyntaxError: future feature annotations is not defined
```

in here:

```
    from pkg_resources.extern import platformdirs
  ...
    from __future__ import annotations
```

The obvious issue is that an incompatible  `pkg_resources` can be in `PYTHONPATH` after `spack env activate`.

E.g. `SPACK_PYTHON` may point to system python while a newer python is in `PATH` for which the modules are in `PYTHONPATH`.

Spack should not have non-vendored 3rd party dependencies, so request to restrict this feature to newer versions of python that have builtin support, or vendor `importlib_metadata` (but rather not...).

**Comment by haampie** [comment]
The other minor point about this PR is that it triggers a system scan on every Spack invocation:

```
$ strace -e file -f python3 -c 'from importlib import metadata; metadata.entry_points()' |& grep entry_points
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/packaging-23.1.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/decorator-5.1.1.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/cycler-0.11.0.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/numpy-1.26.3.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/jmespath-1.0.1.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/kiwisolver-1.4.5.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/isort-5.12.0.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/fonttools-4.39.4.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/s3transfer-0.6.0.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/boto3-1.26.26.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/pickleshare-0.7.5.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/click-8.1.7.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/executing-1.2.0.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/platformdirs-3.10.0.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/mypy_extensions-1.0.0.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/path/to/lib/python3.11/site-packages/vermin-1.6.0.dist-info/entry_points.txt", O_RDONLY|O_CLOEXEC) = 3
...
```

not as bad as Spack's own package cache invalidation, but still.

@tjfulle is there any mechanism in `entry_points.txt` that ensures compatibility with Spack version and Python interpreter? Can it trigger arbitrary function calls?

**Comment by tjfulle** [comment]
@haampie said:

> is there any mechanism in entry_points.txt that ensures compatibility with Spack version and Python interpreter?

`importlib.metadata.entry_points` was introduced in Python 3.8 as a standard way of extending python packages.  It ships with the Python interpreter's standard library, so I am not sure what you mean by ensure compatibility with the interpreter.

@haampie said: 

> Can it trigger arbitrary function calls?

`importlib.metada.entry_points` won't trigger arbitrary function calls.  It will return an entry point that Spack loads.  The loading process will trigger whatever function calls are in the entry point.

**Pull request merged**
